src/lightning_lite/plugins/precision/precision.py

```diff
@@ -0,0 +1,96 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import contextlib
+from typing import Any, Dict, Generator, Optional, Union
+
+from torch import Tensor
+from torch.nn import Module
+from torch.optim import Optimizer
+
+from lightning_lite.utilities.types import _PARAMETERS
+
+
+class PrecisionPlugin:
+    """Base class for all plugins handling the precision-specific parts of the training.
+
+    The class attribute precision must be overwritten in child classes. The default value reflects fp32 training.
+    """
+
+    precision: Union[str, int] = 32
+
+    def get_main_params(self, optimizer: Optimizer) -> _PARAMETERS:
+        """The main params of the model.
+
+        Returns the plain model params here. Maybe different in other precision plugins.
+        """
+        for group in optimizer.param_groups:
+            yield from group["params"]
+
+    @contextlib.contextmanager
+    def forward_context(self) -> Generator[None, None, None]:
+        """A contextmanager for managing model forward/training_step/evaluation_step/predict_step."""
+        yield
+
+    def pre_backward(self, tensor: Tensor, module: Optional[Module]) -> None:
+        """Run before precision plugin executes backward.
+
+        Args:
+            tensor:
+            module:
+        """
+
+    def backward(self, tensor: Tensor, model: Optional[Module], *args: Any, **kwargs: Any) -> None:
+        """Performs the actual backpropagation."""
+        tensor.backward(*args, **kwargs)
+
+    def post_backward(self, tensor: Tensor, module: Optional[Module]) -> None:
```

Comment:
The impl in PL does a `.detach()` here, should we too?

Reply:
I'm mainly unhappy with the fact that we need to return the loss from the call here. That makes it inconsistent with the pre_backward. On top of that, the user would detach the loss themselves in their Lite code. It seems unnecessary to me.

If we add it, it would only exist for PL, but right now it is unclear how we will be able to integrate the Lite precision interface completely (and cleanly) with PL given the mess with apex.

Reply:
> right now it is unclear how we will be able to integrate the Lite precision interface completely (and cleanly) with PL given the mess with apex.

I don't get what would be the problem, but considering #14416, I would not give too much thought into having a nice clean apex plugin

Reply:
Resolving, open to change this once it becomes needed in Lite